bluesky/simulators.py

```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+#plan for later modification.
+_plan_history = {'set':{}, 'trigger':{}}
+
+def est_time(plan, print_output = True):
+    '''The call method.
+
+    This method takes in a plan, and an optional print_output kwarg, and returns an estimate for 
+    the time to complete for the plan, in addition it returns an time estimate for any subplans. 
+
+    PARAMETERS
+    ----------
+    plan : generator.
+        The bluesky plan that the est_time is to be estimated for.
+    print_output : boolean, optional.
+        Indicates if the return values should also be printed to the command line in a human 
+        readable way, default value is 'True'.
+
+    RETURN PARAMETERS
+    -----------------
+    out_time : list.
+        A list containing 2 items, the est_time and the std_dev, for the plan.
+    run_info : list.
+        A list of items, 1 for each run in the plan, containing an est_time/std_dev tuple.
+
+    '''
+    #reset the ._plan_history dictionary. 
+    _plan_history = {'set':{}, 'trigger':{} }
+
+    #Define some variables used in the following.
+    out_time = _TimeStats(0, 0) #this holds the plan est_time and std_dev as a pair.
+    run_info = [] #used to track the ETA and STD_DEV for any 'runs' inside the plan.
+
+    for msg in plan:
+        if msg.command in _run_start_cmds:
+            rn_time = run_est(msg, plan)
+            run_info.append(rn_time)
+            out_time = combine_est(out_time, rn_time)
+        if msg.command in (_group_start_cmds + _flyer_start_cmds):
+            grp_time = group_est(msg, plan)
+            out_time = combine_est(out_time, grp_time)            
+
+        else:
+            object_est = obj_est(msg)
+            out_time = combine_est(out_time, object_est)
+
+        if print_output == True:
+            for i, run in enumerate(run_info):
+                print('  * Run %d est. time --> %.2g s, std Dev --> %.2g s' % (i+1, run[0], run[1]))
+            print('Plan est. time --> %.2g s, std Dev --> %.2g s' % (out_time.est_time, 
+                                                                                out_time.std_dev))
+
+        return out_time, run_info
+
+def msg_est(plan):
+    '''Returns a list of msg/est_time/std_dev named tuples for each msg in plan
+
+    This function is used to return an un-grouped list of msgs and estimated times. It differs 
+    from self.est_time in that it does not estimate the entire plan, or 'runs' in the plan and it 
+    does not sort into msg 'groups', which are done in parallel. it simply returns a msg/est_time/
+    std_dev tuple for each message in the plan.
+
+    PARAMETERS
+    ----------
+    plan : generator.
+        The bluesky plan that the est_time is to be estimated for.
+
+    RETURN PARAMETERS
+    -----------------
+    out_time : list.
+        A list containing a msg/est_time/std_dev named tuple for each message in the plan.
+    '''
+    
+    out_time = []
```

Comment:
Rather than accumulating the messages in a list and then returning the list, this can be implemented as generator, yielding out one estimate at a time. This makes it more flexible. If the caller _wants_ a list, then can of course call `list(msg_est(plan))`.